### ISSUE TITLE
helpers are not available inside the rescue_from block - Why?

### ISSUE THREAD

**Issue by aruprakshit** [ruby-grape/grape#1922]
I have current_user defined inside the before hook, and then we try to read it from `rescue_from` method, we get `nil`.

```ruby
before do
  token         = nil
  token         = headers.fetch("Authorization", "").split("Basic ")[1] unless report_url?
  @current_user = ::User.find_by(auth_token: token) if token.present?

  PaperTrail.request.whodunnit = @current_user.id if @current_user
end

helpers Pundit

helpers do
  attr_reader :current_user

  def report_url?
    request.path_info == "/v1/share_reports/#{request.params['token']}"
  end

  # ....
end


rescue_from Pundit::NotAuthorizedError do |e|
  Airbrake.notify(e, {
    user: {
      id: current_user.id,
      email: current_user.email
    }
  })

  error!(
    {
      messages: {
        userNotification: e.message
      }
    }, 401
  )
end
```

We get `nil` error when `current_user` being accessed inside the `rescue_from` block as `current_user.id`. How can I make available `current_user` inside the `rescue_from` block?

**Comment by dblock** [comment]
I *think* I'd want this to work. Doing `@something = ` in `before` makes it available in the execution of the API, but not in `rescue_from`. Let's call it a bug or a non-feature. Maybe someone could try to fix it?

As a workaround, the rescue block is not executed in the same scope as the request, so you don't have access to attributes. You do have access to `context` (since https://github.com/ruby-grape/grape/pull/1918, in a released version you can reach into ENV), which gives you the endpoint and then gives you endpoint-based context and settings. I think you generally want to do what Warden does, which is using ENV to pass around something across various contexts of the request (before, during, after) and not `@`.

**Comment by dblock** [comment]
Yep. I will leave this open to make `@variable` work. At the very least we should document what the scope of `@` is in a `rescue_from` block. 

@aruprakshit This is a good place for someone to write some tests and contribute :) Wink wink

**Comment by dblock** [comment]
Thanks for the detailed explanation! I don't love copying instance variables at all because modifying them will not work across copies, so I think fixing `self` is the way to go. I understand the other concerns.

**Comment by jcagarcia** [comment]
Totally agree @dblock . I've just tried to copy the instance variables from one instance to another to see if after that they were accessible. 

I've applied the `context` fix solution in #2377
